near-sdk-macros/src/core_impl/contract_metadata/build_info.rs

```diff
@@ -0,0 +1,84 @@
+#[derive(serde::Serialize)]
+pub(crate) struct BuildInfo {
+    build_environment: String,
+    build_command: Vec<String>,
+    contract_path: String,
+    source_code_snapshot: String,
+}
+
+/// basic parsing errors
+#[derive(Debug)]
+pub(super) enum FieldError {
+    EmptyBuildEnvironment,
+    UnsetOrEmptyBuildCommand,
+    /// `None` value should be used instead for `contract_path`
+    UnsetContractPath,
+    UnsetOrEmptySourceSnapshot,
+}
+
+#[allow(clippy::write_literal)]
+impl std::fmt::Display for FieldError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::EmptyBuildEnvironment => {
+                write!(f, "`CARGO_NEAR_BUILD_ENVIRONMENT` is set, but it's set to empty string!")
+            }
+            Self::UnsetOrEmptyBuildCommand => {
+                write!(f, "{}{}",
+                    "`CARGO_NEAR_BUILD_COMMAND` is required, when `CARGO_NEAR_BUILD_ENVIRONMENT` is set, ",
+                    "but it's either not set or empty!"
+                )
+            }
+            Self::UnsetContractPath => {
+                write!(f, "{}{}", "`CARGO_NEAR_CONTRACT_PATH` was provided, ", "but it's not set!")
+            }
+            Self::UnsetOrEmptySourceSnapshot => {
+                write!(f, "{}{}",
+                    "`CARGO_NEAR_SOURCE_CODE_GIT_URL` is required, when `CARGO_NEAR_BUILD_ENVIRONMENT` is set, ",
+                    "but it's either not set or empty!"
+                )
+            }
+        }
+    }
+}
```

Comment:
I don't see a reason to overcomplicate things. Return `Result<Self, String>` from `from_env` and inline the error messages, and call it a day since anyway, you call `to_string` on it to unwrap during the compilation time.

Reply:
k, let me try, i vaguely remember the reason why this seemed like a better way to do this, not sure if it still holds true

Reply:
k, this was done to remove repeating parts of code, and arguably, make it more readable

Reply:
diff in https://github.com/dj8yfo/near-sdk-rs/commit/a34e7ef08f740709b50266768ad2e62993286e6d